File: training-data/csharp/ReverseLinkedList.cs
using System;
using System.Collections.Generic;

class Node
{
    public int Value;
    public Node Next;
    public Node(int value, Node next = null) { Value = value; Next = next; }
}

class ReverseLinkedList
{
    static Node Reverse(Node head)
    {
        Node prev = null;
        while (head != null)
        {
            var next = head.Next;
            head.Next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }

    static List<int> ToList(Node head)
    {
        var result = new List<int>();
        while (head != null)
        {
            result.Add(head.Value);
            head = head.Next;
        }
        return result;
    }

    static void Main()
    {
        var head = new Node(1, new Node(2, new Node(3, new Node(4))));
        Console.WriteLine(string.Join(" ", ToList(Reverse(head))));
    }
}

File: training-data/rust/reverse_linked_list.rs
#[derive(Debug)]
struct Node {
    value: i32,
    next: Option<Box<Node>>,
}

fn reverse(mut head: Option<Box<Node>>) -> Option<Box<Node>> {
    let mut prev = None;
    while let Some(mut node) = head {
        head = node.next.take();
        node.next = prev;
        prev = Some(node);
    }
    prev
}

fn to_vec(mut head: &Option<Box<Node>>) -> Vec<i32> {
    let mut result = vec![];
    while let Some(node) = head {
        result.push(node.value);
        head = &node.next;
    }
    result
}

fn main() {
    let head = Some(Box::new(Node {
        value: 1,
        next: Some(Box::new(Node {
            value: 2,
            next: Some(Box::new(Node { value: 3, next: None })),
        })),
    }));
    println!("{:?}", to_vec(&reverse(head)));
}
